python/src/spark_rapids_ml/umap.py

```diff
@@ -1240,9 +1297,83 @@ def _get_cuml_transform_func(
         Optional[_EvaluateFunc],
     ]:
         cuml_alg_params = self.cuml_params
-        driver_embedding = self.embedding_
-        driver_raw_data = self.raw_data_
-        outputCol = self.getOutputCol()
+        sparse_fit = self._sparse_fit
+        n_cols = self.n_cols
+
+        def _chunk_and_broadcast(
+            sc: pyspark.SparkContext,
+            input: Union[np.ndarray, scipy.sparse.csr_matrix],
+            BROADCAST_LIMIT: int,
+        ) -> Union[
+            List[pyspark.broadcast.Broadcast],
+            Dict[str, List[pyspark.broadcast.Broadcast]],
+        ]:
+            """
+            Broadcast the input array, chunking it into smaller arrays if it exceeds the broadcast limit.
+            """
+            if isinstance(input, np.ndarray):
+                total_bytes = input.nbytes
+                if total_bytes <= BROADCAST_LIMIT:
+                    return [sc.broadcast(input)]
+
+                rows_per_chunk = BROADCAST_LIMIT // (total_bytes // input.shape[0])
+                num_chunks = (input.shape[0] + rows_per_chunk - 1) // rows_per_chunk
+                return [
+                    sc.broadcast(input[i * rows_per_chunk : (i + 1) * rows_per_chunk])
+                    for i in range(num_chunks)
+                ]
+
+            elif isinstance(input, scipy.sparse.csr_matrix):
+                indices, indptr, data = input.indices, input.indptr, input.data
+                total_bytes = sum(arr.nbytes for arr in [indices, indptr, data])
+                if total_bytes <= BROADCAST_LIMIT:
+                    return {
+                        "indices": [sc.broadcast(indices)],
+                        "indptr": [sc.broadcast(indptr)],
+                        "data": [sc.broadcast(data)],
+                    }
+
+                rows_per_chunk = BROADCAST_LIMIT // (total_bytes // data.shape[0])
+                num_chunks = (data.shape[0] + rows_per_chunk - 1) // rows_per_chunk
+                # indptr is chunked differently because it has less rows; split to match num_chunks
+                indptr_rows_per_chunk = (
+                    indptr.shape[0] + (num_chunks - 1)
+                ) // num_chunks
+                return {
+                    "indices": [
+                        sc.broadcast(
+                            indices[i * rows_per_chunk : (i + 1) * rows_per_chunk]
+                        )
+                        for i in range(num_chunks)
+                    ],
```

Comment:
For the second case:
rows_per_chunk = BROADCAST_LIMIT // (total_bytes // len(data)) = BROADCAST_LIMIT // (36 // 4)
=> num_chunks = (4 + 1 - 1) // 1 = 4, so it would work.

For the first case, agreed we definitely need to check the case where data is sufficiently dense that BROADCAST_LIMIT < bytes-per-row. 

In general agree that data skewness is an issue and it's in the things-to-address for 24.12. The current code overestimates the size of each row by using total_bytes // len(data) rather than data.nbytes // len(data) to give some buffer room, but this is kind of arbitrary and needs a revision. 
A simple approach is to check the nbytes of any outgoing chunks to ensure nothing exceeds the limit, and rechunk in a smart way to get under the boundary. Or as Erik mentioned, just reusing the Arrow record batches that were returned from fit, and applying a double-check on the sizes of these. 

test_umap_chunking explicitly tests num_chunks > 1. But certainly it could apply more corner cases, e.g. very large row density and skewness.